test/complexity_test.cc

```diff
@@ -1,37 +1,189 @@
 
-#include "benchmark/benchmark_api.h"
-
-#include <cstdlib>
-#include <string>
+#undef NDEBUG
+#include "benchmark/benchmark.h"
+#include "../src/check.h" // NOTE: check.h is for internal use only!
+#include "../src/re.h"    // NOTE: re.h is for internal use only
+#include <cassert>
+#include <cstring>
+#include <iostream>
+#include <sstream>
 #include <vector>
-#include <map>
+#include <utility>
 #include <algorithm>
 
-std::vector<int> ConstructRandomVector(int size) {
-  std::vector<int> v;
-  v.reserve(size);
-  for (int i = 0; i < size; ++i) {
-    v.push_back(rand() % size);
+namespace {
+
+// ========================================================================= //
+// -------------------------- Testing Case --------------------------------- //
+// ========================================================================= //
+
+enum MatchRules {
+  MR_Default, // Skip non-matching lines until a match is found.
+  MR_Next    // Match must occur on the next line.
+};
+
+struct TestCase {
+  std::string regex;
+  int match_rule;
+
+  TestCase(std::string re, int rule = MR_Default) : regex(re), match_rule(rule) {}
+
+  void Check(std::stringstream& remaining_output) const {
+    benchmark::Regex r;
+    std::string err_str;
+    r.Init(regex, &err_str);
+    CHECK(err_str.empty()) << "Could not construct regex \"" << regex << "\""
+                           << " got Error: " << err_str;
+
+    std::string line;
+    while (remaining_output.eof() == false) {
+        CHECK(remaining_output.good());
+        std::getline(remaining_output, line);
+        if (r.Match(line)) return;
+        CHECK(match_rule != MR_Next) << "Expected line \"" << line
+                                     << "\" to match regex \"" << regex << "\"";
+    }
+
+    CHECK(remaining_output.eof() == false)
+        << "End of output reached before match for regex \"" << regex
+        << "\" was found";
   }
-  return v;
-}
+};
 
-std::map<int, int> ConstructRandomMap(int size) {
-  std::map<int, int> m;
-  for (int i = 0; i < size; ++i) {
-    m.insert(std::make_pair(rand() % size, rand() % size));
+std::vector<TestCase> ConsoleOutputTests;
```

Comment:
Yes, I think it would be a good idea to abstract the implementation of the logic of reporter_output_test.cc, because it is very useful for testing the output of any benchmark on console, json and csv and I think future developments will benefit from it.
At first it seems to me that some things could be moved to a .h file (like a couple of classes and macros), but probably there is a better way to reuse it. I think @EricWF did the reporter_output_test.cc, so may be he can help us on the proper way to restructure it to be reused more efficiently.
@dominichamon : All the other comments are already fixed. Please check if I missed something.